colrpc: enhance warnings from the outbox

yuzefovich opened the pull request:
This commit marks several string constants as "safe" from the
redactability perspective so that the warnings logged by the outboxes
are more helpful. Additionally, several minor nits around error
formatting are addressed.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/68567)
<!-- Reviewable:end -->

knz commented on pkg/sql/colexec/colbuilder/execplan.go:
```diff
@@ -539,9 +540,9 @@ func (r opResult) createAndWrapRowSource(
 		return errors.New("processorConstructor is nil")
 	}
-	log.VEventf(ctx, 1, "planning a row-execution processor in the vectorized flow because %v", causeToWrap)
+	log.VEventf(ctx, 1, "planning a row-execution processor in the vectorized flow because %v", redact.Safe(causeToWrap))
```
Make causeToWrap a redact.RedactableString and prepare redactability at the source.

knz commented on pkg/sql/colexec/colbuilder/execplan.go:
```diff
@@ -539,9 +540,9 @@ func (r opResult) createAndWrapRowSource(
+	log.VEventf(ctx, 1, "planning a row-execution processor in the vectorized flow because %v", redact.Safe(causeToWrap))
 	if err := canWrap(flowCtx.EvalCtx.SessionData.VectorizeMode, spec); err != nil {
-		log.VEventf(ctx, 1, "planning a wrapped processor failed because %v", err)
+		log.VEventf(ctx, 1, "planning a wrapped processor failed because %v", redact.Safe(err))
```
No. Errors are never safe.

knz replied:
Also by convention don't use the word "because" and simply use a colon instead.

knz commented on pkg/sql/colexec/colbuilder/execplan.go:
```diff
@@ -682,7 +683,7 @@ func NewColOperator(
 			result.OpMonitors = result.OpMonitors[:0]
 			if returnedErr != nil {
-				log.VEventf(ctx, 1, "vectorized planning failed with %v", returnedErr)
+				log.VEventf(ctx, 1, "vectorized planning failed with %v", redact.Safe(returnedErr))
```
See above. Also use a colon

knz commented on pkg/sql/colflow/colrpc/inbox.go:
```diff
@@ -279,7 +280,7 @@ func (i *Inbox) Init(ctx context.Context) {
 		// occur. The Inbox must still be closed.
 		i.close()
-		log.VEventf(ctx, 1, "Inbox encountered an error in Init: %v", err)
+		log.VEventf(ctx, 1, "Inbox encountered an error in Init: %v", redact.Safe(err))
```
No.

knz commented on pkg/sql/colflow/colrpc/inbox.go:
```diff
@@ -300,7 +301,7 @@ func (i *Inbox) Next() coldata.Batch {
 			i.close()
 			err := logcrash.PanicAsError(0, panicObj)
-			log.VEventf(i.Ctx, 1, "Inbox encountered an error in Next: %v", err)
+			log.VEventf(i.Ctx, 1, "Inbox encountered an error in Next: %v", redact.Safe(err))
```
No

knz commented on pkg/sql/colflow/colrpc/inbox.go:
```diff
@@ -413,7 +414,7 @@ func (i *Inbox) sendDrainSignal(ctx context.Context) error {
 	if err := i.stream.Send(&execinfrapb.ConsumerSignal{DrainRequest: &execinfrapb.DrainRequest{}}); err != nil {
 		if log.V(1) {
-			log.Warningf(ctx, "Inbox unable to send drain signal to Outbox: %+v", err)
+			log.Warningf(ctx, "Inbox unable to send drain signal to Outbox: %+v", redact.Safe(err))
```
No

knz commented on pkg/sql/colflow/colrpc/inbox.go:
```diff
@@ -443,7 +444,7 @@ func (i *Inbox) DrainMeta() []execinfrapb.ProducerMetadata {
 			}
 			if log.V(1) {
-				log.Warningf(i.Ctx, "Inbox Recv connection error while draining metadata: %+v", err)
+				log.Warningf(i.Ctx, "Inbox Recv connection error while draining metadata: %+v", redact.Safe(err))
```
No

knz commented on pkg/sql/colflow/colrpc/outbox.go:
```diff
@@ -168,7 +169,7 @@ func (o *Outbox) Run(
 	o.runnerCtx = ctx
 	ctx = logtags.AddTag(ctx, "streamID", streamID)
-	log.VEventf(ctx, 2, "Outbox Dialing %s", nodeID)
+	log.VEventf(ctx, 2, "Outbox Dialing %d", redact.SafeInt(nodeID))
```
Not needed. Node IDs are already handled.

knz commented on pkg/sql/colflow/colrpc/outbox.go:
```diff
@@ -177,7 +178,7 @@ func (o *Outbox) Run(
 				ctx,
 				"Outbox Dial connection error, distributed query will fail: %+v",
-				err,
+				redact.Safe(err),
```
No

knz commented on pkg/sql/colflow/colrpc/outbox.go:
```diff
@@ -188,7 +189,7 @@ func (o *Outbox) Run(
 				ctx,
 				"Outbox FlowStream connection error, distributed query will fail: %+v",
-				err,
+				redact.Safe(err),
```
No

knz commented on pkg/sql/colflow/colrpc/outbox.go:
```diff
@@ -201,7 +202,7 @@ func (o *Outbox) Run(
 				ctx,
 				"Outbox Send header error, distributed query will fail: %+v",
-				err,
+				redact.Safe(err),
```
No

knz commented on pkg/sql/colflow/colrpc/outbox.go:
```diff
@@ -226,18 +227,18 @@ func handleStreamErr(
 	if err == io.EOF {
 		if log.V(1) {
-			log.Infof(ctx, "Outbox calling outboxCtxCancel after %s EOF", opName)
+			log.Infof(ctx, "Outbox calling outboxCtxCancel after %s EOF", redact.SafeString(opName))
```
Make opName a safe string or redactable string at the source.

knz commented on pkg/sql/colflow/colrpc/outbox.go:
```diff
@@ -226,18 +227,18 @@ func handleStreamErr(
 		outboxCtxCancel()
 	} else {
-		log.Warningf(ctx, "Outbox calling flowCtxCancel after %s connection error: %+v", opName, err)
+		log.Warningf(ctx, "Outbox calling flowCtxCancel after %s connection error: %+v", redact.SafeString(opName), redact.Safe(err))
```
Ditto opName. Also errors are definitely not safe.

knz commented on pkg/sql/colflow/colrpc/outbox.go:
```diff
@@ -226,18 +227,18 @@ func handleStreamErr(
 func (o *Outbox) moveToDraining(ctx context.Context, reason string) {
 	if atomic.CompareAndSwapUint32(&o.draining, 0, 1) {
-		log.VEventf(ctx, 2, "Outbox moved to draining (%s)", reason)
+		log.VEventf(ctx, 2, "Outbox moved to draining (%s)", redact.SafeString(reason))
```
Make reason a safe or redactable string at the source.

knz commented on pkg/sql/colflow/colrpc/outbox.go:
```diff
@@ -326,7 +327,7 @@ func (o *Outbox) sendBatches(
 	msg := &execinfrapb.ProducerMessage{}
 	if errToSend != nil {
-		log.VEventf(ctx, 1, "Outbox sending an error as metadata: %v", errToSend)
+		log.VEventf(ctx, 1, "Outbox sending an error as metadata: %v", redact.Safe(errToSend))
```
No

knz commented on pkg/sql/colflow/colrpc/outbox.go:
```diff
@@ -395,7 +396,7 @@ func (o *Outbox) runWithStream(
 			switch {
 			case msg.Handshake != nil:
-				log.VEventf(ctx, 2, "Outbox received handshake: %v", msg.Handshake)
+				log.VEventf(ctx, 2, "Outbox received handshake: %v", redact.Safe(msg.Handshake))
```
Make handshake handle the Safe formatter interface instead. It's not generally safe.

knz commented on pkg/sql/colflow/vectorized_flow.go:
```diff
@@ -221,7 +222,7 @@ func (f *vectorizedFlow) Setup(
 		f.creator.cleanup(ctx)
 		f.creator.Release()
-		log.VEventf(ctx, 1, "failed to vectorize: %v", err)
+		log.VEventf(ctx, 1, "failed to vectorize: %v", redact.Safe(err))
```
No

knz commented on pkg/sql/colflow/vectorized_flow.go:
```diff
@@ -273,7 +274,10 @@ func (f *vectorizedFlow) GetPath(ctx context.Context) string {
-	log.VEventf(ctx, 1, "flow %s spilled to disk, stack trace: %s", f.ID, util.GetSmallTrace(2))
+	log.VEventf(
+		ctx, 1, "flow %s spilled to disk, stack trace: %s",
+		redact.SafeString(f.ID.String()), redact.SafeString(util.GetSmallTrace(2)),
```
Don't call String on ID. It's already handled.

I think we already have a redactable variant of the small trace somewhere. If not, you can create it. It's not the responsibility of callers to assert it's safe.

knz commented on pkg/sql/colflow/vectorized_flow.go:
```diff
@@ -273,7 +274,10 @@ func (f *vectorizedFlow) GetPath(ctx context.Context) string {
+		redact.SafeString(f.ID.String()), redact.SafeString(util.GetSmallTrace(2)),
+	)
 	if err := f.Cfg.TempFS.MkdirAll(f.tempStorage.path); err != nil {
 		colexecerror.InternalError(errors.Errorf("unable to create temporary storage directory: %v", err))
```
This needs Wrapf

knz commented on pkg/sql/colflow/vectorized_flow.go:
```diff
@@ -324,9 +328,9 @@ func (f *vectorizedFlow) Cleanup(ctx context.Context) {
-				f.GetID().Short(),
-				f.GetPath(ctx),
-				err,
+				redact.SafeString(f.GetID().Short()),
```
Make Short return a redactable string.

knz commented on pkg/sql/colflow/vectorized_flow.go:
```diff
@@ -324,9 +328,9 @@ func (f *vectorizedFlow) Cleanup(ctx context.Context) {
-				f.GetPath(ctx),
-				err,
+				redact.SafeString(f.GetID().Short()),
+				redact.SafeString(f.GetPath(ctx)),
```
Paths are definitely not safe!!!

knz commented on pkg/sql/colflow/vectorized_flow.go:
```diff
@@ -324,9 +328,9 @@ func (f *vectorizedFlow) Cleanup(ctx context.Context) {
-				err,
+				redact.SafeString(f.GetID().Short()),
+				redact.SafeString(f.GetPath(ctx)),
+				redact.Safe(err),
```
No

knz commented on pkg/sql/colflow/vectorized_flow.go:
```diff
@@ -890,7 +894,7 @@ func (s *vectorizedFlowCreator) setupInput(
 				// diagrams.
 				latency = 0
-				log.VEventf(ctx, 1, "an error occurred during vectorized planning while getting latency: %v", err)
+				log.VEventf(ctx, 1, "an error occurred during vectorized planning while getting latency: %v", redact.Safe(err))
```
No

knz commented on pkg/sql/distsql/server.go:
```diff
@@ -159,7 +159,7 @@ func (ds *ServerImpl) Drain(
 ) {
 	if err := ds.setDraining(true); err != nil {
-		log.Warningf(ctx, "unable to gossip distsql draining state: %s", err)
+		log.Warningf(ctx, "unable to gossip distsql draining state: %s", redact.Safe(err))
```
No

Also use %v not %s

knz commented on pkg/sql/distsql/server.go:
```diff
@@ -227,7 +227,7 @@ func (ds *ServerImpl) setupFlow(
 			req.Version, execinfra.MinAcceptedVersion, execinfra.Version,
 		)
-		log.Warningf(ctx, "%v", err)
+		log.Warningf(ctx, "%v", redact.Safe(err))
```
No

knz commented on pkg/sql/distsql/server.go:
```diff
@@ -528,7 +528,10 @@ func (ds *ServerImpl) SetupLocalSyncFlow(
-	log.VEventf(ctx, 1, "received SetupFlow request from n%v for flow %v", req.Flow.Gateway, req.Flow.FlowID)
+	log.VEventf(
+		ctx, 1, "received SetupFlow request from n%d for flow %s",
+		redact.SafeInt(req.Flow.Gateway), redact.SafeString(req.Flow.FlowID.String()),
```
These types should be handled already. Don't call String on integer-like values.

knz commented on pkg/sql/distsql/server.go:
```diff
@@ -584,7 +587,10 @@ func (ds *ServerImpl) flowStreamInt(
-	log.VEventf(ctx, 1, "connected inbound stream %s/%d", flowID.Short(), streamID)
+	log.VEventf(
+		ctx, 1, "connected inbound stream %s/%d",
+		redact.SafeString(flowID.Short()), redact.SafeInt(streamID),
```
Ditto above

knz commented on pkg/sql/distsql_running.go:
```diff
@@ -299,7 +300,7 @@ func (dsp *DistSQLPlanner) setupFlows(
 		for _, spec := range flows {
 			if err := colflow.IsSupported(vectorizeMode, spec); err != nil {
-				log.VEventf(ctx, 2, "failed to vectorize: %s", err)
+				log.VEventf(ctx, 2, "failed to vectorize: %s", redact.Safe(err))
```
No

knz commented on pkg/sql/distsql_running.go:
```diff
@@ -795,10 +796,10 @@ func (r *DistSQLReceiver) SetError(err error) {
 	// canceled.
 	if r.ctx.Err() != nil {
-		log.VEventf(r.ctx, 1, "encountered error (transitioning to shutting down): %v", r.ctx.Err())
+		log.VEventf(r.ctx, 1, "encountered error (transitioning to shutting down): %v", redact.Safe(r.ctx.Err()))
```
No

knz commented on pkg/sql/distsql_running.go:
```diff
@@ -795,10 +796,10 @@ func (r *DistSQLReceiver) SetError(err error) {
 		r.status = execinfra.ConsumerClosed
 	} else {
-		log.VEventf(r.ctx, 1, "encountered error (transitioning to draining): %v", err)
+		log.VEventf(r.ctx, 1, "encountered error (transitioning to draining): %v", redact.Safe(err))
```
No

knz commented on pkg/sql/exec_util.go:
```diff
@@ -1261,7 +1261,7 @@ func getPlanDistribution(
 	if err != nil {
 		// Don't use distSQL for this request.
-		log.VEventf(ctx, 1, "query not supported for distSQL: %s", err)
+		log.VEventf(ctx, 1, "query not supported for distSQL: %s", redact.Safe(err))
```
No

knz reviewed (commented):
It seems to me that you don't yet fully understand how redactability works. If a type implements redactability already (*especially* if it contains mixes of safe and unsafe information, like errors) then it will expand itself as a mixed redactable string in formatting contexts. You should not use redact.Safe. in many cases in your PR your proposed change is outright invalid, as it promises (via redact.Safe) that you've guaranteed the data "under it" does not contain unsafe information , while errors definitely contain unsafe stuff. in fact, quite generally any use of redact.Safe should be considered code smell.

Use the RedactableString type more liberally, as well as the SafeFormatter interface.

yuzefovich reviewed (commented):
Thanks for taking a look!

Yeah, you're right - I don't truly understand how it all works. One thing I'm confused about is how do I propagate an error as "safe" if I know that it is actually safe? Like I know the full code path an error goes through, and I'm creating the error via `errors.New` without any PII in the first place.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68567#-:-Mgfv6ZiBXboUOYg1-GU:b-slj3ot)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)



<!-- Sent from Reviewable.io -->

yuzefovich committed:
colrpc: enhance warnings from the outbox

This commit marks several string constants as "safe" from the
redactability perspective so that the warnings logged by the outboxes
are more helpful. Additionally, several minor nits around error
formatting are addressed.

Release note: None

yuzefovich reviewed (commented):
I reverted most of the changes since I really wanted to fix the outbox logging. PTAL.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68567#-:-Mgg0WjJCs1VuxaOW7KZ:b9g90ro)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz)

___
*[pkg/sql/colexec/colbuilder/execplan.go, line 543 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68567#-MgSaZgwkHZSOjlSBTLg-r1-543:-Mgg0UkqBIb1u-zrQKcz:bp3aow6) ([raw file](https://github.com/cockroachdb/cockroach/blob/f900db87c96933373cad26190eea5ca84f11fda7/pkg/sql/colexec/colbuilder/execplan.go#L543)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Make causeToWrap a redact.RedactableString and prepare redactability at the source.
</blockquote></details>

Reverted.
___
*[pkg/sql/colexec/colbuilder/execplan.go, line 545 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68567#-MgSaZgwkHZSOjlSBTLg-r1-545:-MgfyuNHE3zQ7eGWKfFq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f900db87c96933373cad26190eea5ca84f11fda7/pkg/sql/colexec/colbuilder/execplan.go#L545)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Also by convention don't use the word "because" and simply use a colon instead.
</blockquote></details>

Done.
___
*[pkg/sql/colexec/colbuilder/execplan.go, line 686 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68567#-MgSaZgwkHZSOjlSBTLg-r1-686:-MgfyvTu0_ffI6gyvJJq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f900db87c96933373cad26190eea5ca84f11fda7/pkg/sql/colexec/colbuilder/execplan.go#L686)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

See above. Also use a colon 
</blockquote></details>

Done.
___
*[pkg/sql/colflow/vectorized_flow.go, line 225 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68567#-MgSaZgx68W-BODcz8-F-r1-225:-Mgfz2XOCq5rRZEXjOi5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f900db87c96933373cad26190eea5ca84f11fda7/pkg/sql/colflow/vectorized_flow.go#L225)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

No
</blockquote></details>

Done.
___
*[pkg/sql/colflow/vectorized_flow.go, line 279 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68567#-MgSaZgx68W-BODcz8-F-r1-279:-MgfzEu67LlRAvtoUwjJ:bp3aow6) ([raw file](https://github.com/cockroachdb/cockroach/blob/f900db87c96933373cad26190eea5ca84f11fda7/pkg/sql/colflow/vectorized_flow.go#L279)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Don't call String on ID. It's already handled.

I think we already have a redactable variant of the small trace somewhere. If not, you can create it. It's not the responsibility of callers to assert it's safe.
</blockquote></details>

Reverted.
___
*[pkg/sql/colflow/vectorized_flow.go, line 282 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68567#-MgSaZgx68W-BODcz8-F-r1-282:-Mgg-_paC6sCwa2lNtMU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f900db87c96933373cad26190eea5ca84f11fda7/pkg/sql/colflow/vectorized_flow.go#L282)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

This needs Wrapf
</blockquote></details>

Done.
___
*[pkg/sql/colflow/vectorized_flow.go, line 331 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68567#-MgSaZgx68W-BODcz8-F-r1-331:-MgfzHX51oanpqO3Lz6C:bp3aow6) ([raw file](https://github.com/cockroachdb/cockroach/blob/f900db87c96933373cad26190eea5ca84f11fda7/pkg/sql/colflow/vectorized_flow.go#L331)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Make Short return a redactable string.
</blockquote></details>

Reverted.
___
*[pkg/sql/colflow/vectorized_flow.go, line 332 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68567#-MgSaZgx68W-BODcz8-F-r1-332:-MgfzInA7GTbMbGXKF3T:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f900db87c96933373cad26190eea5ca84f11fda7/pkg/sql/colflow/vectorized_flow.go#L332)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Paths are definitely not safe!!!
</blockquote></details>

Done.
___
*[pkg/sql/colflow/vectorized_flow.go, line 333 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68567#-MgSaZgx68W-BODcz8-F-r1-333:-MgfzJDJ7DFcPtmnQTlF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f900db87c96933373cad26190eea5ca84f11fda7/pkg/sql/colflow/vectorized_flow.go#L333)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

No
</blockquote></details>

Done.
___
*[pkg/sql/colflow/vectorized_flow.go, line 897 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68567#-MgSaZgx68W-BODcz8-F-r1-897:-MgfzLVs1bpg9sP8JixS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f900db87c96933373cad26190eea5ca84f11fda7/pkg/sql/colflow/vectorized_flow.go#L897)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

No
</blockquote></details>

Done.
___
*[pkg/sql/colflow/colrpc/outbox.go, line 172 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68567#-MgSaZgx68W-BODcz8-E-r1-172:-MgfzSkX7ZLAy9J3_w37:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f900db87c96933373cad26190eea5ca84f11fda7/pkg/sql/colflow/colrpc/outbox.go#L172)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Not needed. Node IDs are already handled.
</blockquote></details>

Done.
___
*[pkg/sql/colflow/colrpc/outbox.go, line 181 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68567#-MgSaZgx68W-BODcz8-E-r1-181:-MgfzU_UDSzLEKnXsxZk:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f900db87c96933373cad26190eea5ca84f11fda7/pkg/sql/colflow/colrpc/outbox.go#L181)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

No
</blockquote></details>

Done.
___
*[pkg/sql/colflow/colrpc/outbox.go, line 192 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68567#-MgSaZgx68W-BODcz8-E-r1-192:-MgfzUz-8SYLcKT83NGT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f900db87c96933373cad26190eea5ca84f11fda7/pkg/sql/colflow/colrpc/outbox.go#L192)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

No
</blockquote></details>

Done.
___
*[pkg/sql/colflow/colrpc/outbox.go, line 205 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68567#-MgSaZgx68W-BODcz8-E-r1-205:-MgfzWV-Bl7U1xgyUDD_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f900db87c96933373cad26190eea5ca84f11fda7/pkg/sql/colflow/colrpc/outbox.go#L205)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

No
</blockquote></details>

Done.
___
*[pkg/sql/colflow/colrpc/outbox.go, line 230 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68567#-MgSaZgx68W-BODcz8-E-r1-230:-Mgfzl1uAc5TNPKqKinJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f900db87c96933373cad26190eea5ca84f11fda7/pkg/sql/colflow/colrpc/outbox.go#L230)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Make opName a safe string or redactable string at the source.
</blockquote></details>

Done.
___
*[pkg/sql/colflow/colrpc/outbox.go, line 234 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68567#-MgSaZgx68W-BODcz8-E-r1-234:-Mgfzljm-r02lsaWWJ7B:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f900db87c96933373cad26190eea5ca84f11fda7/pkg/sql/colflow/colrpc/outbox.go#L234)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Ditto opName. Also errors are definitely not safe.
</blockquote></details>

Done.
___
*[pkg/sql/colflow/colrpc/outbox.go, line 241 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68567#-MgSaZgx68W-BODcz8-E-r1-241:-Mgg-2VUEWNARCpiyWde:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f900db87c96933373cad26190eea5ca84f11fda7/pkg/sql/colflow/colrpc/outbox.go#L241)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Make reason a safe or redactable string at the source.
</blockquote></details>

Done.
___
*[pkg/sql/colflow/colrpc/outbox.go, line 330 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68567#-MgSaZgx68W-BODcz8-E-r1-330:-Mgg-54cCYChTU1gkxUT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f900db87c96933373cad26190eea5ca84f11fda7/pkg/sql/colflow/colrpc/outbox.go#L330)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

No
</blockquote></details>

Done.
___
*[pkg/sql/colflow/colrpc/outbox.go, line 399 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68567#-MgSaZgx68W-BODcz8-E-r1-399:-Mgg-6Yl7fyReHAXZDHf:bp3aow6) ([raw file](https://github.com/cockroachdb/cockroach/blob/f900db87c96933373cad26190eea5ca84f11fda7/pkg/sql/colflow/colrpc/outbox.go#L399)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Make handshake handle the Safe formatter interface instead. It's not generally safe.
</blockquote></details>

Reverted.
___
*[pkg/sql/distsql/server.go, line 162 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68567#-MgSaZgx68W-BODcz8-G-r1-162:-Mgg-G-gDyOxnI4EBHOS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f900db87c96933373cad26190eea5ca84f11fda7/pkg/sql/distsql/server.go#L162)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

No

Also use %v not %s
</blockquote></details>

Done.
___
*[pkg/sql/distsql/server.go, line 230 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68567#-MgSaZgx68W-BODcz8-G-r1-230:-Mgg-IZE2omBtVAwQEvM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f900db87c96933373cad26190eea5ca84f11fda7/pkg/sql/distsql/server.go#L230)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

No
</blockquote></details>

Done.
___
*[pkg/sql/distsql/server.go, line 533 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68567#-MgSaZgx68W-BODcz8-G-r1-533:-Mgg-LE2EIr0Kxtes_aE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f900db87c96933373cad26190eea5ca84f11fda7/pkg/sql/distsql/server.go#L533)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

These types should be handled already. Don't call String on integer-like values.
</blockquote></details>

Done.
___
*[pkg/sql/distsql/server.go, line 592 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68567#-MgSaZgx68W-BODcz8-G-r1-592:-Mgg-Lda34nIqwNKwdnk:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f900db87c96933373cad26190eea5ca84f11fda7/pkg/sql/distsql/server.go#L592)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Ditto above
</blockquote></details>

Done.
___
*[pkg/sql/distsql_running.go, line 303 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68567#-MgSaZgx68W-BODcz8-H-r1-303:-MgfxwZp5uKZLZAQBtmx:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f900db87c96933373cad26190eea5ca84f11fda7/pkg/sql/distsql_running.go#L303)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

No
</blockquote></details>

Done.
___
*[pkg/sql/distsql_running.go, line 799 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68567#-MgSaZgx68W-BODcz8-H-r1-799:-MgfxwxlACbouUfHwDvA:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f900db87c96933373cad26190eea5ca84f11fda7/pkg/sql/distsql_running.go#L799)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

No
</blockquote></details>

Done.
___
*[pkg/sql/distsql_running.go, line 802 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68567#-MgSaZgx68W-BODcz8-H-r1-802:-MgfxxDy30yJmkflFc8O:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f900db87c96933373cad26190eea5ca84f11fda7/pkg/sql/distsql_running.go#L802)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

No
</blockquote></details>

Done.
___
*[pkg/sql/exec_util.go, line 1264 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68567#-MgSaZgx68W-BODcz8-I-r1-1264:-Mgfy2nGDle9AI6iQSKw:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f900db87c96933373cad26190eea5ca84f11fda7/pkg/sql/exec_util.go#L1264)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

No
</blockquote></details>

Done.
___
*[pkg/sql/colflow/colrpc/inbox.go, line 283 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68567#-MgSaZgwkHZSOjlSBTLk-r1-283:-MgfzMbc3FbR50ika6nN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f900db87c96933373cad26190eea5ca84f11fda7/pkg/sql/colflow/colrpc/inbox.go#L283)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

No.
</blockquote></details>

Done.
___
*[pkg/sql/colflow/colrpc/inbox.go, line 304 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68567#-MgSaZgwkHZSOjlSBTLk-r1-304:-MgfzN2y308k9uH1enUa:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f900db87c96933373cad26190eea5ca84f11fda7/pkg/sql/colflow/colrpc/inbox.go#L304)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

No
</blockquote></details>

Done.
___
*[pkg/sql/colflow/colrpc/inbox.go, line 417 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68567#-MgSaZgwkHZSOjlSBTLk-r1-417:-MgfzNWl9ULPn6urj0fE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f900db87c96933373cad26190eea5ca84f11fda7/pkg/sql/colflow/colrpc/inbox.go#L417)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

No
</blockquote></details>

Done.
___
*[pkg/sql/colflow/colrpc/inbox.go, line 447 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68567#-MgSaZgwkHZSOjlSBTLk-r1-447:-MgfzNwf9n5tQGt7qXdZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f900db87c96933373cad26190eea5ca84f11fda7/pkg/sql/colflow/colrpc/inbox.go#L447)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

No
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
>  I'm creating the error via errors.New without any PII in the first place.

Then if you try this experimentally you'll discover that there won't be any redaction markers in the end. It already works! You don't have to do anything extra for this.

(Conversely, if you find a case that doesn't seem to work, we can talk about it together)

One thing that was broken, is that use of `Errorf("... %v", err)` that you've now fixed. This would have broken the recursion. It's possible that you got confused by an error that was flattened by Errorf in that way.

Here are a couple other candidate problems in the colexec sources:

```
pkg/sql/colexec/colbuilder/execplan.go:                                 return r, errors.AssertionFailedf("unexpectedly couldn't plan a cast although IsCastSupported returned true: %v", err)
pkg/sql/colflow/vectorized_flow.go:             colexecerror.InternalError(errors.Errorf("unable to create temporary storage directory: %v", err))
```

One needs `errors.NewAssertionErrorWithWrappedErrf`
the other needs `Wrap`

> Reviewed 3 of 12 files at r1, 9 of 9 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68567#-:-Mgp-wX-B9xYvul3Pbmc:b-jvvnxu)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[pkg/sql/colexec/hash_based_partitioner.go, line 266 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68567#-Mgp-EZmEFA0RbFbkEWV:-Mgp-EZmEFA0RbFbkEWW:b5tklq2) ([raw file](https://github.com/cockroachdb/cockroach/blob/a103de224bb70effa119738db9b1f0d30ece1582/pkg/sql/colexec/hash_based_partitioner.go#L266)):*
> ```Go
> 	op := &hashBasedPartitioner{
> 		unlimitedAllocator: unlimitedAllocator,
> 		name:               redact.SafeString(name),
> ```

Why cast this here? Why not make the function's argument also a redact.SafeString?

(We want to assert throughout that the name variable is not assembled programmatically for example by including a table name)
___
*[pkg/sql/colflow/colrpc/outbox.go, line 330 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68567#-MgSaZgx68W-BODcz8-E-r1-330:-Mgp-ejI2_nHt8rH-gvk:b3p7g2q) ([raw file](https://github.com/cockroachdb/cockroach/blob/f900db87c96933373cad26190eea5ca84f11fda7/pkg/sql/colflow/colrpc/outbox.go#L330)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

still need to revert this
___
*[pkg/sql/colflow/colrpc/outbox.go, line 414 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68567#-Mgp0qzGBEagWgo6tYsR:-Mgp0qzGBEagWgo6tYsS:bdx09o) ([raw file](https://github.com/cockroachdb/cockroach/blob/a103de224bb70effa119738db9b1f0d30ece1582/pkg/sql/colflow/colrpc/outbox.go#L414)):*
> ```Go
> 		reason := redact.SafeString("terminated gracefully")
> 		if errToSend != nil {
> 			reason = "encountered an error when sending batches"
> ```

you've forgotten to include the error.

Try this:

```go
var reason redact.RedactableString
if errToSend != nil {
   reason = redact.Sprintf("encountered error when sending batches: %v", errToSend)
} else {
   reason = redact.Sprint(redact.SafeString("terminated gracefully"))
}
```
___
*[pkg/sql/colflow/colrpc/outbox.go, line 416 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68567#-Mgp17__5OBO2jj-3mim:-Mgp17__5OBO2jj-3min:bacg0fu) ([raw file](https://github.com/cockroachdb/cockroach/blob/a103de224bb70effa119738db9b1f0d30ece1582/pkg/sql/colflow/colrpc/outbox.go#L416)):*
> ```Go
> 			reason = "encountered an error when sending batches"
> 		}
> 		o.moveToDraining(ctx, reason)
> ```

then you'll need to use `redact.RedactableString` in the call interface


<!-- Sent from Reviewable.io -->

knz commented:
see also #68707 FYI

yuzefovich reviewed (commented):
Thanks for the information!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68567#-:-Mh-UFWd7sAVdbKzp89C:b-j94ob7)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz)

___
*[pkg/sql/colexec/hash_based_partitioner.go, line 266 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68567#-Mgp-EZmEFA0RbFbkEWV:-Mh-SrHyB98ZadC_X1CX:b-l97i80) ([raw file](https://github.com/cockroachdb/cockroach/blob/a103de224bb70effa119738db9b1f0d30ece1582/pkg/sql/colexec/hash_based_partitioner.go#L266)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Why cast this here? Why not make the function's argument also a redact.SafeString?

(We want to assert throughout that the name variable is not assembled programmatically for example by including a table name)
</blockquote></details>

Good point, fixed.
___
*[pkg/sql/colflow/colrpc/outbox.go, line 330 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68567#-MgSaZgx68W-BODcz8-E-r1-330:-Mh-T0c-0AphfPv1TKW8:b-834n6i) ([raw file](https://github.com/cockroachdb/cockroach/blob/f900db87c96933373cad26190eea5ca84f11fda7/pkg/sql/colflow/colrpc/outbox.go#L330)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

still need to revert this
</blockquote></details>

Oops.
___
*[pkg/sql/colflow/colrpc/outbox.go, line 414 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68567#-Mgp0qzGBEagWgo6tYsR:-Mh-TTyQ04N8u5sxY7jr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a103de224bb70effa119738db9b1f0d30ece1582/pkg/sql/colflow/colrpc/outbox.go#L414)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

you've forgotten to include the error.

Try this:

```go
var reason redact.RedactableString
if errToSend != nil {
   reason = redact.Sprintf("encountered error when sending batches: %v", errToSend)
} else {
   reason = redact.Sprint(redact.SafeString("terminated gracefully"))
}
```
</blockquote></details>

Done.
___
*[pkg/sql/colflow/colrpc/outbox.go, line 416 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68567#-Mgp17__5OBO2jj-3mim:-Mh-TUD--VHI_zmZoVbv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a103de224bb70effa119738db9b1f0d30ece1582/pkg/sql/colflow/colrpc/outbox.go#L416)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

then you'll need to use `redact.RedactableString` in the call interface
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

knz reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> :100: 

> Reviewed 8 of 8 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68567#-:-MhDkSOWAGB3GbGtmfuh:bbie1ho)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @yuzefovich)



<!-- Sent from Reviewable.io -->

yuzefovich commented:
TFTR!

bors r+
